src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -75,15 +77,22 @@ internal class ProjectBuildManager
             }
         }
 
-        public ProjectBuildManager(ImmutableDictionary<string, string> additionalGlobalProperties, ILogger? msbuildLogger = null)
+        public ProjectBuildManager(ImmutableDictionary<string, string> additionalGlobalProperties, ILogger? msbuildLogger = null, MSB.Execution.BuildManager? msbuildManager = null)
         {
             _additionalGlobalProperties = additionalGlobalProperties ?? ImmutableDictionary<string, string>.Empty;
             _msbuildLogger = msbuildLogger;
+            _msbuildManager = msbuildManager;
+            if (msbuildManager != null)
+            {
+                _msbuildManagerLock = new SemaphoreSlim(initialCount: 1);
+            }
```

Comment:
Could always assign values in the constructor?
```suggestion
            _msbuildManager = msbuildManager ?? MSB.Execution.BuildManager.DefaultBuildManager;
            _msbuildManagerLock = msbuildManager != null ? new SemaphoreSlim(initialCount: 1) : s_buildManagerLock;
```

Reply:
Yes, this makes sense. I will apply your suggestion once CI tests pass once to ensure this change overall does not break anything. Will also need to change usages of `_msbuildManagerLock`.

Reply:
Updated as suggested.

My only concern is that `MSB.Execution.BuildManager.DefaultBuildManager` might be slow to initialize, and if somebody passeses `msbuildManager == null` without ever calling `BuildAsync` or `StartBatchBuild`, they will hit the penalty of that initialization.

Reply:
Hm, either tests are flaky, or initializing these fields in the constructor is indeed problematic.